### PR TITLE
added tests to seperate files

### PR DESCRIPTION
**Pull request by btkostner**
Moved all the tests to respective files in app/tests
Added a changelog test 
Made application status 'FAILED' when needed

### CODE REVIEW

File: app/model/application.js [hunk 1]
```diff
@@ -4,16 +4,18 @@ import Promise from 'bluebird';
 import ini from 'ini';
 import semver from 'semver';
 import _ from 'lodash';
+import dotize from 'dotize';
```
Comment (aroman) [inline app/model/application.js]: TIL this module exists!

File: app/model/application.js [hunk 2]
```diff
@@ -77,74 +68,63 @@ ApplicationSchema.virtual('status').get(function() {
...
-  const application = this;
+// Cleans up application object to something easily loggable
+// Used for development only!
+ApplicationSchema.methods.toClean = function() {
+  let application = _.cloneDeep(this.toObject());
 
-  const i = _.findIndex(application.releases, function(iRelease) {
-    return _.isMatch(iRelease, query)
-  });
+  if (typeof application.icon === 'string') { // Trip super long encoded data
```
Comment (aroman) [inline app/model/application.js]: i see you like using `'string'` instead of `String` (and similar things for other types). Is there a reason, or just personal preference? I tend to think using the actual `Object` is better, personally.

Comment (aroman) [inline app/model/application.js]: also, s/Trip/Trim/ :)

File: app/model/application.js [hunk 3]
```diff
@@ -77,74 +68,63 @@ ApplicationSchema.virtual('status').get(function() {
...
-    $set: {
-      'releases.$': object,
-    },
-  }, (error, info) => {
-    if (error) {
-      return Promise.reject(error)
-    }
-
-    return Application.findOne({_id: application._id});
+  const notDotQuery = _.mapValues(dotQuery, value => {
```
Comment (aroman) [inline app/model/application.js]: could probably one-liner it:

`_.mapValues(dotQuery, value => { $ne: value });`

File: app/model/application.js [hunk 4]
```diff
@@ -77,74 +68,63 @@ ApplicationSchema.virtual('status').get(function() {
...
-      },
-    }, (error, info) => {
-      if (error) {
-        return Promise.reject(error)
-      }
-
-      return Application.findOne({_id: application._id});
-    });
-  }
+  const dotQuery = dotize.convert(query, 'releases');
```
Comment (aroman) [inline app/model/application.js]: could you add a comment to this & `notDotQuery` so it's more clear what these variables are?

File: app/model/application.js [hunk 5]
```diff
@@ -204,130 +181,15 @@ ApplicationSchema.methods.pushLabel = function() {
...
-ApplicationSchema.methods.fetchDesktop = function() {
-  let application = this;
-  const repoName = application.github.name;
-  const fullName = application.github.fullName;
-  app.log.silly(`application fetchDesktop called for ${fullName}`);
-
-  return gh.request(`GET /repos/${fullName}/contents/data/${repoName}.desktop`, {
-    token: application.github.APItoken,
+  .then(data => {
+    return Promise.resolve(null);
```
Comment (aroman) [inline app/model/application.js]: is this a standard idiom for promises?

File: app/model/application.js [hunk 6]
```diff
@@ -362,12 +220,9 @@ ApplicationSchema.methods.pushGithub = function() {
 // Returns saved application including 'errors' always
 ApplicationSchema.methods.syncGithub = function() {
   let application = this;
-  const repoName = application.github.name;
-  const fullName = application.github.fullName;
-  app.log.silly(`application syncGithub called for ${fullName}`);
 
   return application.fetchGithub()
-  .then(application => application.pushGithub()); // Push doesn't save anything
+  .then(application => application.pushGithub); // Push doesn't save anything
```
Comment (aroman) [inline app/model/application.js]: I'm not sure I understand this line — why is `pushGithub` not being called?

File: app/model/application.js [hunk 7]
```diff
@@ -430,13 +288,10 @@ ApplicationSchema.statics.updateBuild = function(data) {
...
     return build.update(data)
-    .catch(error => {
-      return Promise.reject(error);
-    });
   });
 }
 
 // Application creation and exportation
-var Application = mongoose.model('application', ApplicationSchema);
+const Application = mongoose.model('application', ApplicationSchema);
```
Comment (aroman) [inline app/model/application.js]: nice :sunglasses:

File: app/test.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+// Test on update function
+// Import application object
+// Export promise of object with error and warning messages
+export function update(application) {
+  // Array of test functions
+  let loadedTests = [];
+
+  // Push all update functions to loadedTests array
+  for (let file in testFiles) {
```
Comment (aroman) [inline app/test.js]: I'd probably have written this like:

```js
const loadedTests = testFiles
.filter(file => typeof file.update === Function)
.map(file => file.update(application));
```

...but that's just my opinion. I tend to prefer functional/immutable style unless it's really necessary to do something iteratively.

File: app/test.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  app.log.silly(`Running ${loadedTests.length} update tests on ${application.github.fullName}`);
+
+  return Promise.all(loadedTests)
+  .then(messages => { // Consolidate and sort the returned array of objects
+    let props = {
+      error: [],
+      warning: [],
+    }
+
+    for (let i = 0; i < messages.length; i++) {
```
Comment (aroman) [inline app/test.js]: similarly:

```js
const props = {
  error: _.flattenDeep(messages.map(message => error)).sort(),
  warning:_.flattenDeep( messages.map(message => warning)).sort()
}
```

Or actually there's probably a more abstracted way of doing this. Why do you need to `_.flattenDeep`, anyway?

File: app/tests/README.md [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+All tests are expected to be ran in parallel and return a promise of an object in the form of:
+```javaScript
+{
+  error: ['error message strings'],
+  warning: ['warning message strings']
+}
+```
+
+NOTE: If you expect to save data to the database, use the `.update()` function instead of `.save()`. This will prevent data overwrites when running tests in parallel.
```
Comment (aroman) [inline app/tests/README.md]: I'm not sure I understand this — when would a test want/need to save data to the database?

File: app/tests/README.md [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+### Update function
+
+Any file that exports a function named `update` will be ran on repository initialize, and on repository update. It receives an `application` object from the database. All files must be fetched from the GitHub api, as repositories are not cloned unless a release is being tested.
```
Comment (aroman) [inline app/tests/README.md]: I might rephrase this to be a bit more clear, like:

"If the test file exports a function named `update`, it will be called with an `application` object from the database whenever a repository is initialized or updated."

Comment (aroman) [inline app/tests/README.md]: Also: what's the point of having a test file that doesn't export `update`? (asked differently: why not just use `export default`?)

File: app/tests/changelog.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  // Array to store error and warning messages for returning
+  let data = {
+    error: [],
+    warning: [],
+  }
+
+  // Update releases, but keep logic DRY
+  return application.releaseFetchAll()
+  .then(application => {
+    application.releases.forEach(release => {
```
Comment (aroman) [inline app/tests/changelog.js]: might be a bit cleaner split the null check into a separate `filter() step`

File: package.json [hunk 13]
```diff
@@ -17,14 +17,16 @@
     "connect-mongo": "^0.8.1",
     "cookie-parser": "~1.3.4",
     "debug": "~2.1.3",
+    "dotize": "^0.1.26",
     "express": "~4.12.2",
-    "express-handlebars": "~2.0.1",
+    "express-handlebars": "~3.0.0",
```
Comment (aroman) [inline package.json]: anything significantly changed with the major version bump 2->3?

File: views/partials/status/failure.handlebars [hunk 14]
```diff
@@ -1,4 +1,4 @@
-<a href="https://github.com/{{github.fullName}}/issues" target="_blank">
+<a href="https://github.com/{{github.fullName}}/issues?q=is%3Aissue+is%3Aopen+label%3A{{github.label}}" target="_blank">
```
Comment (aroman) [inline views/partials/status/failure.handlebars]: nice!

**Comment by aroman** [comment]
Overall, this looks good! I think you've come up with a nice interface for the testing system. I've commented with my usual minor questions/suggestions/clarifications :)

File: app/model/application.js (hunk 6 above)
Comment (btkostner) [inline app/model/application.js]: that's my bad, I was reading about promises and returning async functions and what not. Must not have changed this back from my time messing around

File: app/test.js (hunk 9 above)
Comment (btkostner) [inline app/test.js]: flattenDeep was more of a just in case a function returns some weird array structure. It really shouldn't be needed

File: app/tests/README.md (hunk 10 above)
Comment (btkostner) [inline app/tests/README.md]: That's mostly for the update functions like the `apphub.js` file. I didn't want to put the fetch and parse logic in the application model, and to keep the code dry, I moved it to the test files. Also, I didn't want to limit the possibilities down the road. The note was more for people that don't understand that the save function will overwrite the current application with the old application object in memory.

use cases:
`apphub.js` saves github label and price to database
`desktop.js` saves application's name
`icon.js` saves the icon

File: app/tests/README.md (hunk 11 above)
Comment (btkostner) [inline app/tests/README.md]: I was planning on added a `pre-build` and `post-build` function for tests at the respective build time.

File: app/tests/changelog.js (hunk 12 above)
Comment (btkostner) [inline app/tests/changelog.js]: something like?
```javascript
  return application.releaseFetchAll()
  .then(application => {
    return Promise.filter(application.releases, release => {
      return release.changelog == null;
    });
  })
  .each(release => {
    data.error.push(`${release.version} ${messages.missing}`);
    return Promise.resolve(null)
  })
```

File: views/partials/status/failure.handlebars (hunk 14 above)
Comment (btkostner) [inline views/partials/status/failure.handlebars]: should be a little easier to find the issues for houston now ;)

File: package.json (hunk 13 above)
Comment (btkostner) [inline package.json]: Slight bump to the original handlebars package, but nothing new. I just like to bump packages I'm working with to keep things up to date. Also gives me slight hope that the problem I'm working with was fixed upstream :P

File: app/test.js (hunk 9 above)
Comment (btkostner) [inline app/test.js]: Considering that's more of a problem on our end, it should never really happen, so I'll clean that up a bit.

File: app/model/application.js (hunk 5 above)
Comment (btkostner) [inline app/model/application.js]: I like to always use the `Promise.resolve` function when returning in promises, but that's mostly personal preference. Considering that a promise always has to return some sort of data (error or not) I just decided to return null. There is no data really worth returning in that function, and I prefer not to return a database lookup of the application to save some io.

File: app/model/application.js (hunk 2 above)
Comment (btkostner) [inline app/model/application.js]: Because `typeof typeof` will always be a string, so it seems more logical to compare it to a string. Most JavaScript code I've seen compares this way.

Comment (aroman) [inline app/model/application.js]: oh whoa I had know idea. Man that's dumb. Anyway yeah, disregard my comment about that then.

File: app/model/application.js (hunk 5 above)
Comment (aroman) [inline app/model/application.js]: sounds good, was just curious to hear your thoughts. Might be worth adding a note about that to the style/coding guidelines ("no-op promises" or something)

File: app/tests/README.md (hunk 10 above)
Comment (aroman) [inline app/tests/README.md]: Oh interesting. So they don't really only test, they load & test. I wonder if test is really the right word to describe them, then. There more "app components" or something. Not necessarily worth holding the PR up, but might be confusing that these aren't strictly test functions, they actually do essential business logic as well.

File: package.json (hunk 13 above)
Comment (aroman) [inline package.json]: yeah for sure, definitely good idea to keep things up to date at this stage!

File: app/test.js (hunk 9 above)
Comment (btkostner) [inline app/test.js]: `_.flattenDeep` clears out any whitespace and blank strings. I think that's why I added it :P

Comment (aroman) [inline app/test.js]: why should there be any empty string/only-whitespace strings in there anyway?

Comment (btkostner) [inline app/test.js]: not sure, the tests just return an empty array, but for some reason when the map function runs, it adds an empty string. I removed the _.flattenDeep and replaced it with a filter instead

File: app/tests/README.md (hunk 10 above)
Comment (btkostner) [inline app/tests/README.md]: True. I didn't want to duplicate the file fetching and parsing from GitHub. I'm not sure what other things we will ever want to save to the database, but it's open in case we want to add things in the future. I'm down for just naming them `components` before we merge.

**Comment by btkostner** [comment]
I renamed the tests to components and uploaded it to a [new branch](https://github.com/elementary/houston/tree/components). Let me know how it looks and i'll merge that one instead.

**Comment by aroman** [comment]
It looks good. It occurs to me though that an even better word might be "hooks", since really that's what you've built, especially if you plan on adding pre- and post- functionality... that really sounds like hook terminology to me. What do you think?

**Comment by btkostner** [comment]
The only gripe I have have about "hooks" is more of a personal one. It sounds very ununique (for lack of a better word) and easy to get mixed up with other hook systems in the application when you are talking about them. How about something like "appHooks"

**Comment by aroman** [comment]
app hooks sounds great, that feels even more accurate/descriptive to me.

I think ultimately we might want to further abstract the app hooks that do tests from the other app hooks, and that would be fine with the terminology you've suggested. (I'm imaging something like TestHook subclass of an AppHook class)
